### DISCUSSION TITLE
Meshing issues

### DISCUSSION THREAD

**Discussion by u01jsrc** [Ifsttar/I-Simpa#364] [Q&A] [answered]
Hi,

I'm using Sketchup (Pro 2023) to create 3D spaces, which we are exporting to 3DS for import into i-Simpa.

However, tetgen.exe in the console is giving this error "Warning:  Point #872 is coincident with #609. Ignored!" for many points.

Having modelled in CATT, Odeon and Raynoise, I've worked hard to try to achieve a "piecewise linear complex" compliant geometry, but still getting the above error when importing the 3DS file.

As a result of the error, i-Simpa refuses to mesh correctly. I can force it via an Average Refresh Model, but am still getting wonky results and would rather fix whatever the issue is so that I am correctly creating the geometry in the first place.

We've been trying this at my consultancy on a few different projects now. With very simple geometry (i.e. 6 face cuboids), we are successful and can fully import and run mesh / calcs in i-Simpa. But as the geometry gets more complicated, the meshing seems to suffer because of it.

Any ideas very welcome (I've read the i-Simpa manual and the TetGen manual, in great depth and spent countless hours with trial-and-error), as we need to be able to reliably model or will have to adopt CATT / Odeon instead (which I want to avoid).

Examples attached in 3DS and STL format.

[Example_geometry.zip](https://github.com/Universite-Gustave-Eiffel/I-Simpa/files/11896591/Example_geometry.zip)

**Comment by wbinek** [comment]
I am not an author of I-SIMPA, but I have used it on several research projects. The model import is usually problematic, as tetgen is very sensitive to geometry errors and there is no real way around it. 

The best workflow I have found is using Blender to fix models, which allows working on individual vertices. In Blender you can enable "Mesh Analysis" overlay and show intersections. Any intersection found in Blender will most likely cause tetgen to fail. Blender also has very usefull mesh clean up options (Edit Mode->Mesh->Clean Up), esspecialy "Merge by distance". Additionally the most common issues I find in my models are duplicate vertices and a case where there is vertex from one surface laying on some edge, but the edge is not split at this point causing an intersection.

The problem with latest Blender versions is that the 3ds import export plugin is no longer supported. I have patched it from some older versions, to make it work with new Blender versions but I don't guarantee it to work (I attatch the plugin below). 
[io_scene_3ds.zip](https://github.com/Universite-Gustave-Eiffel/I-Simpa/files/11898412/io_scene_3ds.zip)

In case of your model I did:
- import 3ds
- Apply all transforms (in object mode: select all objects, Object->Apply->All transforms)
- Merge all objects into one mesh (in object mode: select all objects, Object->Join)
- Show intersections (in edit mode: overlays->Mesh analysis->Intersect) which gives:
![image](https://github.com/Universite-Gustave-Eiffel/I-Simpa/assets/10024390/55e849a9-e40c-4ccf-8dab-c35963d7dd4c)
- Merge by distance (in edit mode: Mesh->Clean Up->Merge by distance) which cleans intersections:
![image](https://github.com/Universite-Gustave-Eiffel/I-Simpa/assets/10024390/726d504e-2541-4aed-83ca-c2fec2c77ec3)
- Split edges by vertices laying on the edges (in edit mode: enable "Auto Merge Vertices" with "Split Edges and Faces" option selected, then in vertex selection mode select all vertices and move them by (0, 0, 0) to enforce edge spliting)
- After all the operations two triangles were missing so they had to be recreated
- Then I-SIMPA claims that there are two intersecting triangle pairs left. I have found at least one of them here, but I have no time now to try and fix it further (the highlighted vertex is detached from the edge underneath it, it may require a bit of manual work):
![image](https://github.com/Universite-Gustave-Eiffel/I-Simpa/assets/10024390/07cd0599-0b9a-4c12-b235-7eaacf563188)

One additional note, merging all objects in blender caused the material information to be lost. In order to preserve it I usally assign different materials in blender and when my mesh is clean I split my mesh by materials.

The model with my modifications (scale may be wrong): 
[test_fix.zip](https://github.com/Universite-Gustave-Eiffel/I-Simpa/files/11898913/test_fix.zip)

[Edit] One more side note - I find it easy to accidently make two vertices in close proximity that are actually not merged into one (are not in exactly the same place) in sketchup. This causes problems with I-SIMPA.

**Reply by u01jsrc** [reply]
Wow, that's an incredibly detailed response, thank you so much. And your time is much appreciated.

I'll give this all a try and report back.

**Reply by nicolas-f** [reply]
Thanks @wbinek for your detailed help.

Good alternative to SketchUP for making a geometry are FreeCAD and OpenSCAD.

FreeCAD has a mesh repair ability but it will destroy non-manifold faces.

**Reply by u01jsrc** [reply]
We're giving the blender option a go now, thanks again. 

I've tried freecad before but not openscad, will try that also.

Do you know if there is a detailed workflow written down on how to successfully create 3D geometry for importing into isimpa? I've done a LOT on other acoustic modelling packages and never had these issues. We're assessing whether we can reliably use isimpa in a professional capacity, alternatively we will have to use CATT / Odeon.

**Reply by nicolas-f** [reply, marked as answer]
From the start we set a constraint to generate 3d mesh of the model, even with the "raytracing" method of SPPS. This is not the case with other room modelling software that accepts even holes in models.

We have created a page for importing 3d model:
https://i-simpa-wiki.readthedocs.io/fr/latest/import_file_recommandations.html

We never found an absolute method to fix all 3d models issues. 

I think the best is to use from the start a 3d modelling software based on constructive solid geomery as OpenSCAD; but it is not as easy to use as an architectural modelling software, and you can't create transmission walls inside the model.
